README.md
[![NPM version](https://badge.fury.io/js/cdk-gitlab-runner.svg)](https://badge.fury.io/js/cdk-gitlab-runner)
[![PyPI version](https://badge.fury.io/py/cdk-gitlab-runner.svg)](https://badge.fury.io/py/cdk-gitlab-runner)
![Release](https://github.com/neilkuan/cdk-gitlab-runner/workflows/Release/badge.svg)

![Downloads](https://img.shields.io/badge/-DOWNLOADS:-brightgreen?color=gray)
![npm](https://img.shields.io/npm/dt/cdk-gitlab-runner?label=npm&color=orange)
![PyPI](https://img.shields.io/pypi/dm/cdk-gitlab-runner?label=pypi&color=blue)

![](https://img.shields.io/badge/iam_role_self-enable-green=?style=plastic&logo=appveyor)
![](https://img.shields.io/badge/vpc_self-enable-green=?style=plastic&logo=appveyor)
![](https://img.shields.io/badge/gitlab_url-customize-green=?style=plastic&logo=appveyor)
![](https://img.shields.io/badge/spotfleet-runner-green=?style=plastic&logo=appveyor)

# Welcome to `cdk-gitlab-runner`

This repository template helps you create gitlab runner on your aws account via AWS CDK one line.

## Note
### Default will help you generate below services:

- VPC
  - Public Subnet (2)
- EC2 (1 T3.micro)

## Before start you need gitlab runner token in your `gitlab project` or `gitlab group`

### In Group

Group > Settings > CI/CD
![group](image/group_runner_page.png)

### In Group

Project > Settings > CI/CD > Runners
![project](image/project_runner_page.png)

## Usage

Replace your gitlab runner token in `$GITLABTOKEN`

### Instance Type

```typescript
import { GitlabContainerRunner } from 'cdk-gitlab-runner';

// If want change instance type to t3.large .
new GitlabContainerRunner(this, 'runner-instance', { gitlabtoken: '$GITLABTOKEN', ec2type:'t3.large' });
// OR
// Just create a gitlab runner , by default instance type is t3.micro .
import { GitlabContainerRunner } from 'cdk-gitlab-runner';

new GitlabContainerRunner(this, 'runner-instance', { gitlabtoken: '$GITLABTOKEN' });})
```

### Gitlab Server Customize Url .

If you want change what you want tag name .

```typescript
// If you want change  what  your self Gitlab Server Url .
import { GitlabContainerRunner } from 'cdk-gitlab-runner';

new GitlabContainerRunner(this, 'runner-instance-change-tag', {
  gitlabtoken: '$GITLABTOKEN',
  gitlaburl: 'https://gitlab.my.com/',
});
```

### Tags

If you want change what you want tag name .

```typescript
// If you want change  what  you want tag name .
import { GitlabContainerRunner } from 'cdk-gitlab-runner';

new GitlabContainerRunner(this, 'runner-instance-change-tag', {
  gitlabtoken: '$GITLABTOKEN',
  tags: ['aa', 'bb', 'cc'],
});
```

### IAM Policy

If you want add runner other IAM Policy like s3-readonly-access.

```typescript
// If you want add runner other IAM Policy like s3-readonly-access.
import { GitlabContainerRunner } from 'cdk-gitlab-runner';
import { ManagedPolicy } from '@aws-cdk/aws-iam';

const runner = new GitlabContainerRunner(this, 'runner-instance-add-policy', {
  gitlabtoken: '$GITLABTOKEN',
  tags: ['aa', 'bb', 'cc'],
});
runner.runnerRole.addManagedPolicy(
  ManagedPolicy.fromAwsManagedPolicyName('AmazonS3ReadOnlyAccess'),
);
```

### Security Group

If you want add runner other SG Ingress .

```typescript
// If you want add runner other SG Ingress .
import { GitlabContainerRunner } from 'cdk-gitlab-runner';
import { Port, Peer } from '@aws-cdk/aws-ec2';

const runner = new GitlabContainerRunner(this, 'runner-add-SG-ingress', {
  gitlabtoken: 'GITLABTOKEN',
  tags: ['aa', 'bb', 'cc'],
});

// you can add ingress in your runner SG .
runner.defaultRunnerSG.connections.allowFrom(
  Peer.ipv4('0.0.0.0/0'),
  Port.tcp(80),
);
```

### Use self VPC

> 2020/06/27 , you can use your self exist VPC or new VPC , but please check your `vpc public Subnet` Auto-assign public IPv4 address must be Yes ,or `vpc private Subnet` route table associated `nat gateway` .

```typescript
import { GitlabContainerRunner } from 'cdk-gitlab-runner';
import { Port, Peer, Vpc, SubnetType } from '@aws-cdk/aws-ec2';
import { ManagedPolicy } from '@aws-cdk/aws-iam';

const newvpc = new Vpc(stack, 'VPC', {
  cidr: '10.1.0.0/16',
  maxAzs: 2,
  subnetConfiguration: [
    {
      cidrMask: 26,
      name: 'RunnerVPC',
      subnetType: SubnetType.PUBLIC,
    },
  ],
  natGateways: 0,
});

const runner = new GitlabContainerRunner(this, 'testing', {
  gitlabtoken: '$GITLABTOKEN',
  ec2type: 't3.small',
  selfvpc: newvpc,
});
```

### Use your self exist role

> 2020/06/27 , you can use your self exist role assign to runner

```typescript
import { GitlabContainerRunner } from 'cdk-gitlab-runner';
import { Port, Peer } from '@aws-cdk/aws-ec2';
import { ManagedPolicy, Role, ServicePrincipal } from '@aws-cdk/aws-iam';

const role = new Role(this, 'runner-role', {
  assumedBy: new ServicePrincipal('ec2.amazonaws.com'),
  description: 'For Gitlab EC2 Runner Test Role',
  roleName: 'TestRole',
});

const runner = new GitlabContainerRunner(stack, 'testing', {
  gitlabtoken: '$GITLAB_TOKEN',
  ec2iamrole: role,
});
runner.runnerRole.addManagedPolicy(
  ManagedPolicy.fromAwsManagedPolicyName('AmazonS3ReadOnlyAccess'),
);
```

### Custom Gitlab Runner EBS szie

> 2020/08/22 , you can change you want ebs size.

```typescript
import { GitlabContainerRunner } from 'cdk-gitlab-runner';

new GitlabContainerRunner(stack, 'testing', {
  gitlabtoken: '$GITLAB_TOKEN',
  ebsSize: 50,
});
```

### Control the number of runners with AutoScalingGroup

> 2020/11/25 , you can set the number of runners.

```typescript
import { GitlabRunnerAutoscaling } from 'cdk-gitlab-runner';

new GitlabRunnerAutoscaling(stack, 'testing', {
  gitlabToken: '$GITLAB_TOKEN',
  minCapacity: 2,
  maxCapacity: 2,
});
```

### Support Spotfleet Gitlab Runner

> 2020/08/27 , you can use spotfleet instance be your gitlab runner,
> after create spotfleet instance will auto output instance id .thank [@pahud](https://github.com/pahud/cdk-spot-one) again ~~~

```typescript
import { GitlabContainerRunner, BlockDuration } from 'cdk-gitlab-runner';

const runner = new GitlabContainerRunner(stack, 'testing', {
  gitlabtoken: 'GITLAB_TOKEN',
  ec2type: 't3.large',
  blockDuration: BlockDuration.ONE_HOUR,
  spotFleet: true,
});
// configure the expiration after 1 hours
runner.expireAfter(Duration.hours(1));
```

> 2020/11/19, you setting job runtime bind host volumes.
> see more https://docs.gitlab.com/runner/configuration/advanced-configuration.html#the-runnersdocker-section

```typescript
import { GitlabContainerRunner, BlockDuration } from 'cdk-gitlab-runner';

const runner = new GitlabContainerRunner(stack, 'testing', {
  gitlabtoken: 'GITLAB_TOKEN',
  ec2type: 't3.large',
  dockerVolumes: [
    {
      hostPath: '/tmp/cahce',
      containerPath: '/tmp/cahce',
    },
  ],
});
```
> 2020/11/19, support runner auto unregister runner when cdk app destroy.

# Note

![](https://img.shields.io/badge/version-1.47.1-green=?style=plastic&logo=appveyor) vs ![](https://img.shields.io/badge/version-1.49.1-green=?style=plastic&logo=appveyor)

> About change instance type

This is before ![](https://img.shields.io/badge/version-1.47.1-green=?style) ( included ![](https://img.shields.io/badge/version-1.47.1-green=?style) )

```typescript
import { InstanceType, InstanceClass, InstanceSize } from '@aws-cdk/aws-ec2';
import { GitlabContainerRunner } from 'cdk-gitlab-runner';

// If want change instance type to t3.large .
new GitlabContainerRunner(this, 'runner-instance', {
  gitlabtoken: '$GITLABTOKEN',
  ec2type: InstanceType.of(InstanceClass.T3, InstanceSize.LARGE),
});
```

This is ![](https://img.shields.io/badge/version-1.49.1-green=?style)

```typescript
import { GitlabContainerRunner } from 'cdk-gitlab-runner';

// If want change instance type to t3.large .
new GitlabContainerRunner(this, 'runner-instance', {
  gitlabtoken: '$GITLABTOKEN',
  ec2type: 't3.large',
});
```

## Wait about 6 mins , If success you will see your runner in that page .

![runner](image/group_runner2.png)

#### you can use tag `gitlab` , `runner` , `awscdk` ,

## Example _`gitlab-ci.yaml`_

[gitlab docs see more ...](https://docs.gitlab.com/ee/ci/yaml/README.html)

```yaml
dockerjob:
  image: docker:18.09-dind
  variables:

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.94.19](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.94.18...v1.94.19) (2021-04-21)

### [1.94.18](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.94.17...v1.94.18) (2021-04-20)

### [1.94.17](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.94.16...v1.94.17) (2021-04-19)

### [1.94.16](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.94.15...v1.94.16) (2021-04-18)

### [1.94.15](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.94.14...v1.94.15) (2021-04-17)

### [1.94.14](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.94.13...v1.94.14) (2021-04-16)

### [1.94.13](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.94.12...v1.94.13) (2021-04-15)

### [1.94.12](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.94.11...v1.94.12) (2021-04-14)

### [1.94.11](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.94.10...v1.94.11) (2021-04-13)

### [1.94.10](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.94.9...v1.94.10) (2021-04-12)

### [1.94.9](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.94.8...v1.94.9) (2021-04-11)

### [1.94.8](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.94.7...v1.94.8) (2021-04-10)

### [1.94.7](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.94.6...v1.94.7) (2021-04-09)

### [1.94.6](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.94.5...v1.94.6) (2021-04-08)

### [1.94.5](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.94.4...v1.94.5) (2021-04-07)

### [1.94.4](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.94.3...v1.94.4) (2021-04-06)

### [1.94.3](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.94.2...v1.94.3) (2021-04-05)

### [1.94.2](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.94.1...v1.94.2) (2021-03-29)

### [1.94.1](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.90.20...v1.94.1) (2021-03-21)

### [1.90.20](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.90.19...v1.90.20) (2021-03-17)

### [1.90.19](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.90.18...v1.90.19) (2021-03-16)

### [1.90.18](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.90.17...v1.90.18) (2021-03-09)

### [1.90.17](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.90.16...v1.90.17) (2021-03-08)

### [1.90.16](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.90.15...v1.90.16) (2021-03-07)

### [1.90.15](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.90.14...v1.90.15) (2021-03-06)

### [1.90.14](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.90.13...v1.90.14) (2021-03-05)

### [1.90.13](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.90.12...v1.90.13) (2021-03-04)

### [1.90.12](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.90.11...v1.90.12) (2021-03-03)

### [1.90.11](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.90.10...v1.90.11) (2021-03-02)

### [1.90.10](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.90.9...v1.90.10) (2021-03-01)

### [1.90.9](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.90.8...v1.90.9) (2021-02-28)

### [1.90.8](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.90.7...v1.90.8) (2021-02-27)

### [1.90.7](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.90.6...v1.90.7) (2021-02-26)

### [1.90.6](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.90.5...v1.90.6) (2021-02-25)

### [1.90.5](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.90.4...v1.90.5) (2021-02-24)

### [1.90.4](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.90.3...v1.90.4) (2021-02-23)

### [1.90.3](https://github.com/guan840912/cdk-gitlab-runner/compare/v1.90.2...v1.90.3) (2021-02-22)


### Bug Fixes


src/gitlab-runner-instance.ts
import * as path from 'path';
import {
  Instance,
  InstanceType,
  MachineImage,
  UserData,
  BlockDeviceVolume,
  AmazonLinuxGeneration,
  SubnetType,
  Vpc,
  IVpc,
  IInstance,
  SecurityGroup,
  Port,
  CfnLaunchTemplate,
  CfnSpotFleet,
  ISecurityGroup,
  SubnetSelection,
} from '@aws-cdk/aws-ec2';
import {
  IRole,
  Role,
  ServicePrincipal,
  ManagedPolicy,
  CfnInstanceProfile,
  PolicyStatement,
} from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import * as logs from '@aws-cdk/aws-logs';
import { Bucket } from '@aws-cdk/aws-s3';
import {
  Construct,
  CfnOutput,
  Fn,
  Stack,
  Duration,
  Lazy,
  CustomResource,
  Token,
  RemovalPolicy,
} from '@aws-cdk/core';
import * as cr from '@aws-cdk/custom-resources';

import { DockerVolumes } from './gitlab-runner-interfaces';
export interface GitlabContainerRunnerProps {
  /**
   * Gitlab token for the Register Runner .
   *
   * @example
   * new GitlabContainerRunner(stack, 'runner', { gitlabtoken: 'GITLAB_TOKEN' });
   *
   * @default - You must to give the token !!!
   *
   */
  readonly gitlabtoken: string;

  /**
   * Image URL of Gitlab Runner.
   *
   * @example
   * new GitlabRunnerAutoscaling(stack, 'runner', { gitlabToken: 'GITLAB_TOKEN', gitlabRunnerImage: 'gitlab/gitlab-runner:alpine' });
   *
   * @default public.ecr.aws/gitlab/gitlab-runner:alpine
   *
   */
  readonly gitlabRunnerImage?: string;

  /**
   * Runner default EC2 instance type.
   *
   * @example
   * new GitlabContainerRunner(stack, 'runner', { gitlabtoken: 'GITLAB_TOKEN', ec2type: 't3.small' });
   *
   * @default - t3.micro
   *
   */
  readonly ec2type?: string;

  /**
   * VPC for the Gitlab Runner .
   *
   * @example
   * const newvpc = new Vpc(stack, 'NEWVPC', {
   *   cidr: '10.1.0.0/16',
   *   maxAzs: 2,
   *   subnetConfiguration: [{
   *     cidrMask: 26,
   *     name: 'RunnerVPC',
   *     subnetType: SubnetType.PUBLIC,
   *   }],
   *   natGateways: 0,
   * });
   *
   * new GitlabContainerRunner(stack, 'runner', { gitlabtoken: 'GITLAB_TOKEN', selfvpc: newvpc });
   *
   * @default - new VPC will be created , 1 Vpc , 2 Public Subnet .
   *
   */
  readonly selfvpc?: IVpc;

  /**
   * IAM role for the Gitlab Runner Instance .
   *
   * @example
   * const role = new Role(stack, 'runner-role', {
   *   assumedBy: new ServicePrincipal('ec2.amazonaws.com'),
   *   description: 'For Gitlab EC2 Runner Test Role',
   *   roleName: 'Myself-Runner-Role',
   * });
   *
   * new GitlabContainerRunner(stack, 'runner', { gitlabtoken: 'GITLAB_TOKEN', ec2iamrole: role });
   *
   * @default - new Role for Gitlab Runner Instance , attach AmazonSSMManagedInstanceCore Policy .
   *
   */
  readonly ec2iamrole?: IRole;

  /**
   * Gitlab Runner register tag1  .
   *
   * @example
   * new GitlabContainerRunner(stack, 'runner', { gitlabtoken: 'GITLAB_TOKEN', tag1: 'aa' });
   *
   * @deprecated - use tags ['runner', 'gitlab', 'awscdk']
   *
   * @default - tag1: gitlab .
   *
   */
  readonly tag1?: string;

  /**
   * Gitlab Runner register tag2  .
   *
   * @example
   * new GitlabContainerRunner(stack, 'runner', { gitlabtoken: 'GITLAB_TOKEN', tag2: 'bb' });
   *
   * @deprecated - use tags ['runner', 'gitlab', 'awscdk']
   *
   * @default - tag2: awscdk .
   *
   */
  readonly tag2?: string;

  /**
   * Gitlab Runner register tag3  .
   *
   * @example
   * new GitlabContainerRunner(stack, 'runner', { gitlabtoken: 'GITLAB_TOKEN', tag3: 'cc' });
   *
   * @deprecated - use tags ['runner', 'gitlab', 'awscdk']
   *
   * @default - tag3: runner .
   *
   */
  readonly tag3?: string;

  /**
   * tags for the runner
   *
   * @default - ['runner', 'gitlab', 'awscdk']
   */
  readonly tags?: string[];

  /**
   * Gitlab Runner register url .
   *
   * @example
   * const runner = new GitlabContainerRunner(stack, 'runner', { gitlabtoken: 'GITLAB_TOKEN',gitlaburl: 'https://gitlab.com/'});
   *
   * @default - gitlaburl='https://gitlab.com/' , please use https://yourgitlab.com/ do not use https://yourgitlab.com
   *
   */
  readonly gitlaburl?: string;

  /**
   * Gitlab Runner instance EBS size .
   *
   * @example
   * const runner = new GitlabContainerRunner(stack, 'runner', { gitlabtoken: 'GITLAB_TOKEN',ebsSize: 100});
   *
   * @default - ebsSize=60
   *
   */
  readonly ebsSize?: number;

  /**
   * Gitlab Runner instance Use Spot Fleet or not ?!.
   *
   * @example
   * const runner = new GitlabContainerRunner(stack, 'runner', { gitlabtoken: 'GITLAB_TOKEN',spotFleet: true});
   *
   * @default - spotFleet=false
   *
   */
  readonly spotFleet?: boolean;

  /**
   * SSH key name
   *
   * @default - no ssh key will be assigned , !!! only support spotfleet runner !!! .
   */
  readonly keyName?: string;

  /**
   * Reservce the Spot Runner instance as spot block with defined duration
   *
   * @default - BlockDuration.ONE_HOUR , !!! only support spotfleet runner !!! .
   */
  readonly blockDuration?: BlockDuration;

  /**
   * The behavior when a Spot Runner Instance is interrupted
   *
   * @default - InstanceInterruptionBehavior.TERMINATE , !!! only support spotfleet runner !!! .
   */
  readonly instanceInterruptionBehavior?: InstanceInterruptionBehavior;

  /**
   * the time when the spot fleet allocation expires
   *
   * @default - no expiration , !!! only support spotfleet runner !!! .
   */
  readonly validUntil?: string;

  /**
   * VPC subnet for the spot fleet
   *
   * @example
   * const vpc = new Vpc(stack, 'nat', {
   * natGateways: 1,
   * maxAzs: 2,
   * });
   * const runner = new GitlabContainerRunner(stack, 'testing', {
   *   gitlabtoken: 'GITLAB_TOKEN',
   *   ec2type: 't3.large',
   *   ec2iamrole: role,
   *   ebsSize: 100,
   *   selfvpc: vpc,
   *   vpcSubnet: {
   *     subnetType: SubnetType.PUBLIC,
   *   },
   * });
   *
   * @default - public subnet
   */
  readonly vpcSubnet?: SubnetSelection;

  /**
   * add another Gitlab Container Runner Docker Volumes Path at job runner runtime.
   *
   * more detail see https://docs.gitlab.com/runner/configuration/advanced-configuration.html#the-runnersdocker-section
   *
   * @default - already mount "/var/run/docker.sock:/var/run/docker.sock"
   *
   * @example
   * dockerVolumes: [
   *   {
   *     hostPath: '/tmp/cache',
   *     containerPath: '/tmp/cache',
   *   },
   * ],
   */
  readonly dockerVolumes?: DockerVolumes[];
}

export enum BlockDuration {
  ONE_HOUR = 60,
  TWO_HOURS = 120,
  THREE_HOURS = 180,
  FOUR_HOURS = 240,
  FIVE_HOURS = 300,
  SIX_HOURS = 360,
  SEVEN_HOURS = 420,
  EIGHT_HOURS = 480,
  NINE_HOURS = 540,
  TEN_HOURS = 600,
  ELEVEN_HOURS = 660,
  TWELVE_HOURS = 720,
  THIRTEEN_HOURS = 780,
  FOURTEEN_HOURS = 840,
  FIFTEEN_HOURS = 900,
  SIXTEEN_HOURS = 960,
  SEVENTEEN_HOURS = 1020,
  EIGHTTEEN_HOURS = 1080,
  NINETEEN_HOURS = 1140,
  TWENTY_HOURS = 1200,
}

export enum InstanceInterruptionBehavior {
  HIBERNATE = 'hibernate',
  STOP = 'stop',
  TERMINATE = 'terminate',
}

export class GitlabContainerRunner extends Construct {
  /**
   * The IAM role assumed by the Runner instance .
   */
  public readonly runnerRole: IRole;

  /**
   * This represents a Runner EC2 instance , !!! only support On-demand runner instance !!!
   */
  public readonly runnerEc2!: IInstance;

  /**
   * The EC2 runner's vpc.
   */
  public readonly vpc!: IVpc;

  /**
   * The time when the the fleet allocation will expire , !!! only support spotfleet runner !!!
   */
  private validUntil?: string;

  /**
   * The EC2 runner's default SecurityGroup.
   */
  public readonly defaultRunnerSG!: ISecurityGroup;

  /**
   * SpotFleetRequestId for this spot fleet , !!! only support spotfleet runner !!!
   */
  public readonly spotFleetRequestId!: string;

  /**
   * the first instance id in this fleet , !!! only support spotfleet runner !!!
   */
  readonly spotFleetInstanceId!: string;

  constructor(scope: Construct, id: string, props: GitlabContainerRunnerProps) {
    super(scope, id);
    const spotFleetId = id;

    const defaultProps = {
      gitlabRunnerImage: 'public.ecr.aws/gitlab/gitlab-runner:alpine',
      gitlaburl: 'https://gitlab.com/',
      ec2type: 't3.micro',
      tags: ['gitlab', 'awscdk', 'runner'],
    };
    const runnerProps = { ...defaultProps, ...props };

    const runnerBucket = new Bucket(this, 'runnerBucket', {
      removalPolicy: RemovalPolicy.DESTROY,
    });
    const shell = UserData.forLinux();
    shell.addCommands(...this.createUserData(runnerProps, runnerBucket.bucketName));

    this.runnerRole =
      runnerProps.ec2iamrole ??
      new Role(this, 'runner-role', {
        assumedBy: new ServicePrincipal('ec2.amazonaws.com'),
        description: 'For Gitlab EC2 Runner Role',
      });
    this.validUntil = runnerProps.validUntil;
    const instanceProfile = new CfnInstanceProfile(this, 'InstanceProfile', {
      roles: [this.runnerRole.roleName],
    });
    runnerBucket.grantWrite(this.runnerRole);
    this.vpc =
      runnerProps.selfvpc ??
      new Vpc(this, 'VPC', {
        cidr: '10.0.0.0/16',
        maxAzs: 2,
        subnetConfiguration: [
          {
            cidrMask: 26,
            name: 'RunnerVPC',
            subnetType: SubnetType.PUBLIC,
          },
        ],
        natGateways: 0,
      });
    this.defaultRunnerSG = new SecurityGroup(this, 'SpotFleetSg', {
      vpc: this.vpc,
    });
    this.defaultRunnerSG.connections.allowFromAnyIpv4(Port.tcp(22));
    const spotOrOnDemand = runnerProps.spotFleet ?? false;
    if (spotOrOnDemand) {
      //throw new Error('yes new spotfleet');

      const imageId = MachineImage.latestAmazonLinux({
        generation: AmazonLinuxGeneration.AMAZON_LINUX_2,
      }).getImage(this).imageId;
      const lt = new CfnLaunchTemplate(this, 'LaunchTemplate', {
        launchTemplateData: {
          imageId,
          instanceType: runnerProps.ec2type,
          blockDeviceMappings: [
            {
              deviceName: '/dev/xvda',
              ebs: {
                volumeSize: runnerProps.ebsSize ?? 60,
              },
            },
          ],
          userData: Fn.base64(shell.render()),
          keyName: runnerProps.keyName,
          tagSpecifications: [
            {
              resourceType: 'instance',
              tags: [
                {
                  key: 'Name',
                  value: `${Stack.of(this).stackName
                  }/spotFleetGitlabRunner/${spotFleetId}`,
                },
              ],
            },
          ],
          instanceMarketOptions: {
            marketType: 'spot',
            spotOptions: {
              blockDurationMinutes:
                runnerProps.blockDuration ?? BlockDuration.ONE_HOUR,
              instanceInterruptionBehavior:
                runnerProps.instanceInterruptionBehavior ??
                InstanceInterruptionBehavior.TERMINATE,
            },
          },
          securityGroupIds: this.defaultRunnerSG.connections.securityGroups.map(
            (m) => m.securityGroupId,
          ),
          iamInstanceProfile: {
            arn: instanceProfile.attrArn,
          },
        },
      });

      const spotFleetRole = new Role(this, 'FleetRole', {
        assumedBy: new ServicePrincipal('spotfleet.amazonaws.com'),
        managedPolicies: [
          ManagedPolicy.fromAwsManagedPolicyName(
            'service-role/AmazonEC2SpotFleetTaggingRole',
          ),
        ],
      });

      const vpcSubnetSelection = runnerProps.vpcSubnet ?? {
        subnetType: SubnetType.PUBLIC,
      };
      const subnetConfig = this.vpc
        .selectSubnets(vpcSubnetSelection)
        .subnets.map((s) => ({
          subnetId: s.subnetId,
        }));

      const cfnSpotFleet = new CfnSpotFleet(this, id, {
        spotFleetRequestConfigData: {
          launchTemplateConfigs: [
            {
              launchTemplateSpecification: {
                launchTemplateId: lt.ref,
                version: lt.attrLatestVersionNumber,
              },
              overrides: subnetConfig,
            },
          ],
          iamFleetRole: spotFleetRole.roleArn,
          targetCapacity: 1,
          validUntil: Lazy.stringValue({ produce: () => this.validUntil }),
          terminateInstancesWithExpiration: true,
        },
      });
      const onEvent = new lambda.Function(this, 'OnEvent', {
        code: lambda.Code.fromAsset(path.join(__dirname, '../assets/functions')),
        handler: 'index.on_event',
        runtime: lambda.Runtime.PYTHON_3_8,
        timeout: Duration.seconds(60),
      });

      const isComplete = new lambda.Function(this, 'IsComplete', {
        code: lambda.Code.fromAsset(path.join(__dirname, '../assets/functions')),
        handler: 'index.is_complete',
        runtime: lambda.Runtime.PYTHON_3_8,
        timeout: Duration.seconds(60),
        role: onEvent.role,
      });

      const myProvider = new cr.Provider(this, 'MyProvider', {
        onEventHandler: onEvent,
        isCompleteHandler: isComplete,
        logRetention: logs.RetentionDays.ONE_DAY,
      });

      onEvent.addToRolePolicy(
        new PolicyStatement({
          actions: ['ec2:DescribeSpotFleetInstances'],
          resources: ['*'],
        }),
      );

      const fleetInstances = new CustomResource(this, 'GetInstanceId', {
        serviceToken: myProvider.serviceToken,
        properties: {
          SpotFleetRequestId: cfnSpotFleet.ref,
        },
      });

      fleetInstances.node.addDependency(cfnSpotFleet);
      this.spotFleetInstanceId = Token.asString(
        fleetInstances.getAtt('InstanceId'),
      );
      this.spotFleetRequestId = Token.asString(
        fleetInstances.getAtt('SpotInstanceRequestId'),
      );
      new CfnOutput(this, 'InstanceId', { value: this.spotFleetInstanceId });
      new CfnOutput(this, 'SpotFleetId', { value: cfnSpotFleet.ref });
    } else {
      this.runnerEc2 = new Instance(this, 'GitlabRunner', {
        instanceType: new InstanceType(runnerProps.ec2type),
        instanceName: 'Gitlab-Runner',
        vpc: this.vpc,
        vpcSubnets: runnerProps.vpcSubnet ?? {
          subnetType: SubnetType.PUBLIC,
        },
        machineImage: MachineImage.latestAmazonLinux({
          generation: AmazonLinuxGeneration.AMAZON_LINUX_2,
        }),
        role: this.runnerRole,
        userData: shell,
        securityGroup: this.defaultRunnerSG,
        blockDevices: [
          {
            deviceName: '/dev/xvda',
            volume: BlockDeviceVolume.ebs(runnerProps.ebsSize ?? 60),
          },
        ],
      });
      new CfnOutput(this, 'Runner-Instance-ID', {
        value: this.runnerEc2.instanceId,
      });
    }

    const unregisterRunnerOnEvent = new lambda.Function(this, 'unregisterRunnerOnEvent', {
      code: lambda.Code.fromAsset(path.join(__dirname, '../assets/functions')),
      handler: 'unregister_runner.on_event',
      runtime: lambda.Runtime.PYTHON_3_8,
      timeout: Duration.seconds(60),
    });

    const unregisterRunnerProvider = new cr.Provider(this, 'unregisterRunnerProvider', {
      onEventHandler: unregisterRunnerOnEvent,
      logRetention: logs.RetentionDays.ONE_DAY,
    });

    const unregisterRunnerCR = new CustomResource(this, 'unregisterRunnerCR', {
      resourceType: 'Custom::unregisterRunnerProvider',
      serviceToken: unregisterRunnerProvider.serviceToken,
      properties: {
        BucketName: runnerBucket.bucketName,
        GitlabUrl: runnerProps.gitlaburl,
      },
    });

    runnerBucket.grantReadWrite(unregisterRunnerOnEvent);
    unregisterRunnerCR.node.addDependency(runnerBucket);
    this.runnerRole.addManagedPolicy(
      ManagedPolicy.fromAwsManagedPolicyName('AmazonSSMManagedInstanceCore'),
    );

    new CfnOutput(this, 'Runner-Role-Arn', {
      value: this.runnerRole.roleArn,
    });
  }

  /**
   * @default - !!! only support spotfleet runner !!! .
   */
  public expireAfter(duration: Duration) {
    const date = new Date();
    date.setSeconds(date.getSeconds() + duration.toSeconds());
    this.validUntil = date.toISOString();
  }

  private dockerVolumesList(dockerVolume: DockerVolumes[] | undefined): string {
    let tempString: string = '--docker-volumes "/var/run/docker.sock:/var/run/docker.sock"';
    if (dockerVolume) {
      let tempList: string[] = [];
      dockerVolume.forEach(e => {
        tempList.push(`"${e.hostPath}:${e.containerPath}"`);
      });
      tempList.forEach(e => {
        tempString = `${tempString} --docker-volumes ${e}`;
      });
    }
    return tempString;
  }

  public createUserData(props: GitlabContainerRunnerProps, bucketName: string): string[] {
    return [
      'yum update -y ',
      'sleep 15 && amazon-linux-extras install docker && yum install -y amazon-cloudwatch-agent && systemctl start docker && usermod -aG docker ec2-user && chmod 777 /var/run/docker.sock',
      'systemctl restart docker && systemctl enable docker',
      `docker run -d -v /home/ec2-user/.gitlab-runner:/etc/gitlab-runner -v /var/run/docker.sock:/var/run/docker.sock \
      --name gitlab-runner-register ${props.gitlabRunnerImage} register --non-interactive --url ${props.gitlaburl} --registration-token ${props.gitlabtoken} \
      --docker-pull-policy if-not-present ${this.dockerVolumesList(props?.dockerVolumes)} \
      --executor docker --docker-image "alpine:latest" --description "Docker Runner" \
      --tag-list "${props.tags?.join(',')}" --docker-privileged`,
      `sleep 2 && docker run --restart always -d -v /home/ec2-user/.gitlab-runner:/etc/gitlab-runner -v /var/run/docker.sock:/var/run/docker.sock --name gitlab-runner ${props.gitlabRunnerImage}`,
      `TOKEN=$(cat /home/ec2-user/.gitlab-runner/config.toml | grep token | cut -d '"' -f 2) && echo '{"token": "TOKEN"}' > /tmp/runnertoken.txt && sed -i s/TOKEN/$TOKEN/g /tmp/runnertoken.txt && aws s3 cp /tmp/runnertoken.txt s3://${bucketName}`,
    ];
  }
}
